Semalab/agent
# Add support for Dependency-Check

# Add support for Dependency-Check

## Details
[Dependency-Check](https://github.com/jeremylong/DependencyCheck) can now be run through the `--dependency-check` flag or `AGENT_DEPENDENCY_CHECK` environment variable.

The generated zip now has a `.sema/dependency-check/` directory with the `csv` contents. For example, running a scan on the agent repo produces a zip with the following contents
```
Archive:  test.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  05-17-2023 13:56   .sema/
     7318  05-17-2023 13:54   poetry.lock
        0  05-17-2023 13:56   .sema/dependency_check/
      374  05-17-2023 13:56   .sema/dependency_check/dependency-check-report.csv
---------                     -------
     7692                     4 files
```

### Specific Changes
1. `--dependency-check` or `AGENT_DEPENDENCY_CHECK` must point to the path where Dependency-Check [release](https://github.com/jeremylong/DependencyCheck/releases/download/v8.2.1/dependency-check-8.2.1-release.zip) was downloaded and unzipped.
2. The linux `Dockerfile.run` has been updated to include dependency-check and it's prerequisites, and therefore can be used for testing.
3. The windows `Dockerfile.run` has __not__ been updated to include dependency-check, as it's going to take a little longer for me to figure out how to install some JDK under wine.

## Testing
To test this, a few options:
1. Use the linux docker runner. It already includes the `AGENT_DEPENDENCY_CHECK` environment variable pointing to a valid dependency-check install, and has JDK installed. The `run-bin` script has `/src` pointing to the repository root, so this will run a scan on the `agent` repository itself:
  ```sh
  # from the repository root:
  make linux
  ./scripts/run-bin.sh linux /src --output /src/test.zip
  ```
2. Run it through python locally. You must install dependency-check and java, and then do:
  ```sh
  poetry run cli <repository> --output <output.zip> --dependency-check <path to dependency-check unzipped directory>
  ```
3. Run it as a binary locally. You must install dependency-check and java, and then do:
  ```sh
  # windows
  make windows
  ./dist/windows/main.exe <repository> --output <output.zip> --dependency-cehck <path to dependency-check unzipped directory>

  # linux
  make linux
  ./dist/linux/main <repository> --output <output.zip> --dependency-cehck <path to dependency-check unzipped directory>
  ```

# Easily Testing on a Custom Repo

If you want to scan a specific repository without installing anything you can do the following:
```
make linux

export REPOSITORY=<your repository>
export OUTPUT=<output directory>

docker build -t agent-linux-runner -f ./docker/linux/Dockerfile.run ./
docker run \
  --mount type=bind,source="$PWD"/dist,target=/dist \
  --mount type=bind,source="$REPOSITORY",target=/src \
  --mount type=bind,source="$OUTPUT",target=/output \
  agent-linux-runner \
  /dist/linux/main /src --output /output/output.zip
```
Running this from the repository root will let you scan a separate repository. This sets up the appropriate mounts so the docker linux runner can run the executable, read the repo, and output the zip file.
